### ISSUE TITLE
SWC Plugin

### ISSUE THREAD

**Issue by stnwk** [blitz-js/superjson#157]
`Next.js` seems to be switching from `babel` to `swc`, see: https://github.com/vercel/next.js/pull/29811

In the next stable release, `swc` will be used by default to compile a `Next.js` project if there is no custom `babel.config.js`, with the goal to eventually completely drop support for `babel`.

I guess this means, there will be a need for a `superjson-swc` plugin of some sort? `babel-plugin-macro` is having a similar discussion about it here: https://github.com/kentcdodds/babel-plugin-macros/issues/144

What are your thoughts?

**Comment by flybayer** [comment]
@Skn0tt you can already use the Rust plugin APIs. Next.js is using them. 

Here's the source for the next.js plugins: https://github.com/vercel/next.js/tree/canary/packages/next/build/swc/src

**Comment by flybayer** [comment]
I managed to get more information from the Vercel team: 
https://twitter.com/flybayer/status/1454547398521393153

TLDR: There are no docs for SWC plugins because there is not yet an official SWC Rust plugin API. 

The SWC Rust plugin API is starting to be worked on now.

So not much we can do for now.

**Comment by elie222** [comment]
`warn  - Disabled SWC as replacement for Babel because of custom Babel configuration ".babelrc" https://nextjs.org/docs/messages/swc-disabled`

A quick fix would just be to manually use this package when needed instead of relying on Babel to do it, right?

**Comment by flybayer** [comment]
@elie222 correct

**Comment by elie222** [comment]
Ah, just tried it: `superjson.serialize().json` means you lose the TS types. will pass for now :(

**Comment by CanKattwinkel** [comment]
> Ah, just tried it: `superjson.serialize().json` means you lose the TS types. will pass for now :(

Then you might want to wrap all your props simply with 

```
type Serialized<T> = {
    [P in keyof T]: T[P] extends Date ? string : Serialized<T[P]>
};
``` 

**Disclaimer:** this might need to be extended to cover any object structure, js data type.

**Comment by remorses** [comment]
I made a nextjs plugin that does not disable the SWC transform

You can try it [here](https://github.com/remorses/next-superjson)

Basically it applies the superjson babel plugin only on the pages files, without adding a .babelrc file

**Comment by Skn0tt** [comment]
Putting this here for anybody who comes across this issue: @remorses's plugin, while not adding a `.babelrc` file, still [uses Babel](https://github.com/remorses/next-superjson/blob/959b3dc6063976ac2298a86b7169e38034b7bcbe/next-superjson/src/index.ts#L25). I consider this to be a work-around, the real goal will be to create an `SWC` equivalent of `babel-plugin-superjson-next`.

**Comment by elie222** [comment]
Does it matter if it’s using Babel if SWC works? Or are we still seeing slow builds because of it. I’ve started using remorses package.

Best,
Eliezer

**Comment by Skn0tt** [comment]
It will introduce another bundler into the pipeline, which will slow things down, yes. It won't be as big of a slowdown as the full-on babel strategy which happens when there's a `.babelrc` (this will fully disable SWC, and use Babel for everything), but I suspect it's noticeably slower. It's also somewhat of a weird setup, where code goes through two transpilers (first babel, then SWC); and I suspect that this is not officially unsupported by Next.js. Maybe @timneutkens can shed some light.

**Comment by remorses** [comment]
@Skn0tt the Babel plugin only runs on the pages directory, this can’t be a performance problem unless you have a page 10k lines long

**Comment by Skn0tt** [comment]
That's a good point @remorses! Thank you, I missed that.

**Comment by cacode-euan** [comment]
Any updates on this?

**Comment by kevinsimper** [comment]
Could this be solved by using Next.js new middleware? That seems like a possible solution, adding superjson to all responses and then if one wants to use superjson in the component, they need to call superjson.parse?

**Comment by markhughes** [comment]
I would be happy to try make this into a plugin? It's a fairly simple plugin to make judging by the babel one

**Comment by Skn0tt** [comment]
That would be great @markhughes!

**Comment by markhughes** [comment]
I've been able to write the base of it but next.js does not yet allow adding custom swc plugins. Once they update swc and add plugin support I'll give it a proper test.

**Comment by cryptoAlgorithm** [comment]
@elie222 maybe using the stringify and parse methods can help preserve types? From a quick look at the readme it seems like sending the stringified string as a prop then parsing it with a type will preserve the type.

**Comment by johnmasonwilliams** [comment]
Any update on this?

**Comment by markhughes** [comment]
I've been meaning to test it in the nextjs experimental stuff :) i'll put it up on github i've just been super sick.

**Comment by orionmiz** [comment]
I'm working on swc plugin of superjson with some supports from main creator of swc.

Also, a stale [issue](https://github.com/blitz-js/babel-plugin-superjson-next/issues/121) in current babel plugin would be resolved too.

Coming ASAP!

**Comment by orionmiz** [comment]
Done! You've been waiting for a long time.

🚀 [**next-superjson-plugin**](https://github.com/orionmiz/next-superjson-plugin)

Please check README.md and follow the instructions to apply the plugin.

I've already tested working in Next.js 12.2.4

But if you find any bugs, please leave issues and I will check it out quickly.

**Comment by flybayer** [comment]
@orionmiz that's awesome, great work!!!
